### PR TITLE
Add the ability to preview viewports from the inspector

### LINKED ISSUE #2471: Add inspector UI that lets you spy on what a viewport is rendering (ViewportSpy)
### Describe the project you are working on
This proposal is useful for anyone juggling many Viewports in their projects, and is relatively simple; adding some UI to the editor inspector for Viewports that would let you view what it's currently rendering.

### Describe the problem or limitation you are having in your project
For my case I'm working on [Cartographer](https://github.com/awkwardpolygons/cartographer), a terrain editor and renderer, that uses many Viewports for painting and rendering, I've ended up constantly creating throwaway Sprites or TextureRects to peek at what a Viewport is doing when trying to debug a problem, prototype an idea, or check Viewport settings.

I'm sure it would also be useful for scenarios where Viewports/Cameras are being used to simply render different parts of a scene and you'd like to check them as you work.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
It would be a very simple addition to the inspector when a Viewport node is selected, adding a TextureRect or similar that would show what the Viewport is currently rendering. No longer requiring throwaway nodes to peek on a Viewport.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
I've already implemented it as a plugin, here: [ViewportSpy](https://github.com/awkwardpolygons/viewport-spy)

### If this enhancement will not be used often, can it be worked around with a few lines of script?
A few lines of script? No. A relatively simple plugin? Yes.

### Is there a reason why this should be core and not an add-on in the asset library?
It is an add-on available in the [asset library](https://godotengine.org/asset-library/asset/892), but several folks on [Reddit](https://www.reddit.com/r/godot/comments/m5cpjv/viewportspy_an_inspector_tool_that_lets_you_spy/?utm_source=share&utm_medium=web2x&context=3) and [Github](https://github.com/awkwardpolygons/viewport-spy/issues/1) requested I submit this proposal so that everyone can have access to this simple feature by default.

### PR DESCRIPTION
**Pull request by AaronRecord**
Resolves https://github.com/godotengine/godot-proposals/issues/2471

https://user-images.githubusercontent.com/50304111/124334225-aad84200-db53-11eb-8b41-bd9744179c4d.mp4

### CODE REVIEW

**Comment by LightningAA** [comment]
Would it make sense for this plugin to inherit from `TextureEditorPlugin`? I'm thinking it might since they do pretty much the same thing, but I'll probably need to refactor `TextureEditorPlugin`.

**Comment by LightningAA** [comment]
Alright, I refactored TextureEditorPlugin so that I could reuse the same code for ViewportSpy

**Comment by LightningAA** [comment]
Okay, I got it to pass checks. I haven't tested on 4.0 locally since my laptop doesn't support vulkan, but I have tested on the 3.x branch and everything seems to work fine.

**Comment by KoBeWi** [comment]
ViewportSpy sounds ok for a plugin from AssetLib, but not necessarily a built-in plugin. I'd rename it to ViewportPreview (or something similar).

**Comment by LightningAA** [comment]
> ViewportSpy sounds ok for a plugin from AssetLib, but not necessarily a built-in plugin. I'd rename it to ViewportPreview (or something similar).

Agreed, I'll rename it

File: editor/plugins/texture_editor_plugin.cpp [hunk 1]
```diff
@@ -30,132 +30,60 @@
...
-		//get_scene()->connect("node_removed",this,"_node_removed");
-	}
-
-	if (p_what == NOTIFICATION_DRAW) {
-		Ref<Texture2D> checkerboard = get_theme_icon("Checkerboard", "EditorIcons");
-		Size2 size = get_size();
+	TextureRect *checkerboard = memnew(TextureRect);
+	checkerboard->set_texture(theme->get_icon("Checkerboard", "EditorIcons"));
+	checkerboard->set_stretch_mode(TextureRect::STRETCH_TILE);
+	checkerboard->set_custom_minimum_size(Size2(0.0, 256.0));
```
Comment (Calinou) [inline editor/plugins/texture_editor_plugin.cpp:39]: Scale for hiDPI displays:

```suggestion
	checkerboard->set_custom_minimum_size(Size2(0.0, 256.0) * EDSCALE);
```

File: editor/plugins/texture_editor_plugin.cpp [hunk 2]
```diff
@@ -30,132 +30,60 @@
...
+		metadata_label->set_text(itos(p_texture->get_width()) + "x" + itos(p_texture->get_height()) + " " + format);
 
-void TextureEditor::_texture_changed() {
-	if (!is_visible()) {
-		return;
-	}
-	update();
-}
+		Ref<Font> metadata_label_font = theme->get_font("expression", "EditorFonts");
+		metadata_label->add_theme_font_size_override("font_size", 32);
```
Comment (Calinou) [inline editor/plugins/texture_editor_plugin.cpp:64]: Scale for hiDPI displays:

```suggestion
		metadata_label->add_theme_font_size_override("font_size", 32 * EDSCALE);
```

**Review by Calinou — CHANGES_REQUESTED** [review]

File: editor/plugins/texture_editor_plugin.cpp (hunk 2 above)
Comment (LightningAA) [reply editor/plugins/texture_editor_plugin.cpp]: @Calinou fixed.

File: editor/plugins/texture_editor_plugin.cpp [hunk 3]
```diff
@@ -29,133 +29,62 @@
...
-		//get_scene()->connect("node_removed",this,"_node_removed");
-	}
-
-	if (p_what == NOTIFICATION_DRAW) {
-		Ref<Texture2D> checkerboard = get_theme_icon("Checkerboard", "EditorIcons");
-		Size2 size = get_size();
+	TextureRect *checkerboard = memnew(TextureRect);
+	checkerboard->set_texture(theme->get_icon("Checkerboard", "EditorIcons"));
+	checkerboard->set_stretch_mode(TextureRect::STRETCH_TILE);
+	checkerboard->set_custom_minimum_size(Size2(0.0, 256.0) * EDSCALE);
```
Comment (KoBeWi) [inline editor/plugins/texture_editor_plugin.cpp:40]: Adding `checkerboard->set_texture_repeat(CanvasItem::TEXTURE_REPEAT_ENABLED);` will fix the weird stretching bug.

**Comment by KoBeWi** [comment]
Seems like the preview doesn't update aspect ratio properly. You need to re-select the node to fix it:
6eTfigkH3o

**Comment by KoBeWi** [comment]
This label should have some outline
image
It's potentially unreadable on bright textures.

File: editor/plugins/viewport_preview_editor_plugin.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "viewport_preview_editor_plugin.h"
+
+bool EditorInspectorPluginViewportPreview::can_handle(Object *p_object) {
+	return Object::cast_to<Viewport>(p_object) != nullptr;
```
Comment (KoBeWi) [inline editor/plugins/viewport_preview_editor_plugin.cpp:34]: Should be probably SubViewport, otherwise it appears for all dialog types etc.

Comment (LightningAA) [reply editor/plugins/viewport_preview_editor_plugin.cpp:34]: Should the plugin maybe be called `SubViewportPreview` instead?

Comment (KoBeWi) [reply editor/plugins/viewport_preview_editor_plugin.cpp:34]: Yep, it would be better.

File: editor/plugins/texture_editor_plugin.cpp (hunk 3 above)
Comment (LightningAA) [reply editor/plugins/texture_editor_plugin.cpp:40]: This didn't happen when I tested on 3.x (my laptop doesn't support vulkan). It seems a little weird to me that setting the stretch mode to tile wouldn't actually cause it to tile, should `CanvasItem::set_texture_repeat()` maybe be called from `TextureRect::set_stretch_mode()`? Or maybe `TextureRect::STRETCH_TILE` should just be removed altogether in favor of using `CanvasItem::TEXTURE_REPEAT_ENABLED` + `TexureRect::STRETCH_SCALE` (does that even work?)

**Comment by LightningAA** [comment]
> Seems like the preview doesn't update aspect ratio properly. You need to re-select the node to fix it.

@KoBeWi 
Yeah, I was wondering how to solve this. Do you mean something like connecting [`Viewport.size_changed()`](https://docs.godotengine.org/en/stable/classes/class_viewport.html#signals) to `texture_display->grab_focus()`? Or should I focus the entire `TexturePreview`? Or did you mean something else?

File: editor/plugins/viewport_preview_editor_plugin.cpp (hunk 4 above)
Comment (LightningAA) [reply editor/plugins/viewport_preview_editor_plugin.cpp]: Done.

**Comment by KoBeWi** [comment]
Ah no, by "You need to re-select the node to fix it." I meant that it's the wrong behavior, as seen in the GIF. The preview should properly refresh the size ratio.

**Comment by LightningAA** [comment]
Okay, I get what you mean now. I'll try updating the texture in `size_changed()` and see if that fixes it.

**Comment by LightningAA** [comment]
Okay I think I fixed the resizing issue.

File: editor/plugins/texture_editor_plugin.cpp (hunk 3 above)
Comment (KoBeWi) [reply editor/plugins/texture_editor_plugin.cpp:45]: STRETCH_SCALE is something completely different, we can't have a flag that has a different behavior based on texture settings. It's counter-intuitive.

Although the fact that STRETCH_TILE doesn't work without repeat sounds like a bug 🤔 but it makes sense, old code probably used some custom drawing to make sure it works with any texture and that's less efficient.

**Comment by LightningAA** [comment]
> This label should have some outline

Fixed, I think (someone who's computer can run Godot 4.0 will have to check).

**Comment by KoBeWi** [comment]
You still need to fix the checkerboard. Also there's a compilation bug in `sub_viewport_preview_editor_plugin.cpp` (remove `()` from `::update()`).

**Comment by LightningAA** [comment]
> Also there's a compilation bug in `sub_viewport_preview_editor_plugin.cpp` (remove `()` from `::update()`).

Thanks, I was trying to figure out what that was. I added the fix you suggested to the checkerboard.

**Review by KoBeWi — APPROVED** [review]

**Comment by LightningAA** [comment]
Rebase following https://github.com/godotengine/godot/pull/49948

**Comment by LightningAA** [comment]
@Calinou I've made the changes you've requested, but because I amended my previous commit (instead of commiting your suggestions and then squashing the commits) it still says that you're requesting changes and merging is blocked. I'm not sure how much that matters to akien/what difference it makes, but if you have a minute could you double check that everything is good? 🙂

**Review by Calinou — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**